apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx

```diff
@@ -27,12 +29,85 @@ interface Props {
   type: EditorType;
 }
 
-const transformTreeData = (node: TreeNode): TreeNode => ({
-  ...node,
-  children: node.children?.length
-    ? node.children.map(transformTreeData)
-    : undefined,
-});
+const removeEmptyChildren = (node: TreeNode): TreeNode => {
+  const transformedChildren =
+    node.children
+      ?.map(removeEmptyChildren)
+      .filter((child) => child.children?.length || !child.children) || [];
+
+  return {
+    ...node,
+    children: transformedChildren.length > 0 ? transformedChildren : undefined,
+  };
+};
+
+const parseNodeData =
+  (node: TreeNode, identifier: string) => (param, index) => ({
+    id: `${node.id}-${identifier}-${index}`,
+    actualNode: param,
+    label: param.type,
+    start: param.start,
+    end: param.end,
+  });
+const getPosition = (node: TreeNode) => `${node.start}-${node.end}`;
+const transformTreeData = (node: TreeNode, relation?: string): TreeNode => {
+  const result: TreeNode = {
+    ...node,
+    relation,
+    children: [],
+  };
+
+  if (node.actualNode.type === "CallExpression") {
+    const args = node.actualNode.arguments.map(parseNodeData(node, "arg"));
+    const argumentsIds = args.map(getPosition);
+    result.children =
+      node.children.map((child) => {
+        const isParam = argumentsIds.includes(getPosition(child));
+        return transformTreeData(child, isParam ? "argument" : child.label);
+      }) || [];
+  } else if (node.actualNode.type === "FunctionDeclaration") {
+    const params = node.actualNode.params.map(parseNodeData(node, "param"));
+    const paramsIds = params.map(getPosition);
+
+    // Transform remaining children
+    result.children =
+      node.children.map((child) => {
+        const isParam = paramsIds.includes(getPosition(child));
+        const nameMap = {
+          BlockStatement: "body",
+          Identifier: "id",
+        };
+        return transformTreeData(
+          child,
+          (isParam ? "param" : nameMap[child.label]) || child.label,
+        );
+      }) || [];
+  } else {
```

Comment:
prefer early returns over nested conditionals if it's possible to use them please

Reply:
same as above - we dont have nested conditions - we need to check the node type to make correct mapping (and add relations)